firedrake/assemble.py

```diff
@@ -1185,23 +1184,21 @@ def _apply_bc(self, tensor, bc):
         elif isinstance(bc, EquationBCSplit):
             bc.zero(tensor)
             type(self)(bc.f, bcs=bc.bcs, form_compiler_parameters=self._form_compiler_params, needs_zeroing=False,
-                       zero_bc_nodes=self._zero_bc_nodes, diagonal=self._diagonal).assemble(tensor=tensor)
+                       zero_bc_nodes=self._zero_bc_nodes, diagonal=self._diagonal, weight=self._weight).assemble(tensor=tensor)
         else:
             raise AssertionError
 
     def _apply_dirichlet_bc(self, tensor, bc):
-        if not self._zero_bc_nodes:
-            tensor_func = tensor.riesz_representation(riesz_map="l2")
-            if self._diagonal:
-                bc.set(tensor_func, 1)
-            else:
-                bc.apply(tensor_func)
-            tensor.assign(tensor_func.riesz_representation(riesz_map="l2"))
+        if self._diagonal:
+            bc.set(tensor, self._weight)
+        elif not self._zero_bc_nodes:
+            # We cannot set primal data on a dual Cofunction, this will throw an error
```

Comment:
I don't understand this comment. Will the line below crash?

Reply:
It will crash on `assemble(form, bcs=bcs, zero_bc_nodes=False)`. I think it only makes sense to apply bcs on things like `assemble(interpolate(expr, V), bcs=bcs)`, but I believe that this case is not supported.

Reply:
Maybe this could be rephrased to something like
```
# NOTE: This will only work for assembling XXX because YYY.
```

My complaint is just that the comment is unclear.